kernel/mempool.c

```diff
@@ -0,0 +1,404 @@
+ * "partners" in the bottom 4 bits of the return value
+ */
+static int partner_bits(struct k_mem_pool *p, int level, int bn)
+{
+	u32_t *word;
+	int bit = get_bit_ptr(p, level, bn, &word);
+
+	return (*word >> (4*(bit / 4))) & 0xf;
+}
+
+static size_t buf_size(struct k_mem_pool *p)
+{
+	return p->n_max * p->max_sz;
+}
+
+static bool block_fits(struct k_mem_pool *p, void *block, size_t bsz)
+{
+	return (block + bsz - 1 - p->buf) < buf_size(p);
+}
+
+static void init_mem_pool(struct k_mem_pool *p)
+{
+	int i;
+	size_t buflen = p->n_max * p->max_sz, sz = p->max_sz;
+	u32_t *bits = p->buf + buflen;
+
+	sys_dlist_init(&p->wait_q);
+
+	for (i = 0; i < p->n_levels; i++) {
+		int nblocks = buflen / sz;
+
+		sys_dlist_init(&p->levels[i].free_list);
+
+		if (nblocks < 32) {
+			p->max_inline_level = i;
+		} else {
+			p->levels[i].bits_p = bits;
+			bits += (nblocks + 31)/32;
+		}
+
+		sz = _ALIGN4(sz / 4);
+	}
+
+	for (i = 0; i < p->n_max; i++) {
+		void *block = block_ptr(p, p->max_sz, i);
+
+		sys_dlist_append(&p->levels[0].free_list, block);
+		set_free_bit(p, 0, i);
+	}
+}
+
+int init_static_pools(struct device *unused)
+{
+	ARG_UNUSED(unused);
+	struct k_mem_pool *p;
+
+	for (p = _k_mem_pool_list_start; p < _k_mem_pool_list_end; p++) {
+		init_mem_pool(p);
+	}
+
+	return 0;
+}
+
+SYS_INIT(init_static_pools, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
+
+/* A note on synchronization: all manipulation of the actual pool data
+ * happens in one of alloc_block()/free_block() or break_block().  All
+ * of these transition between a state where the caller "holds" a
+ * block pointer that is marked used in the store and one where she
+ * doesn't (or else they will fail, e.g. if there isn't a free block).
+ * So that is the basic operation that needs synchronization, which we
+ * can do piecewise as needed in small one-block chunks to preserve
+ * latency.  At most (in free_block) a single locked operation
+ * consists of four bit sets and dlist removals. If the overall
+ * allocation operation fails, we just free the block we have (putting
+ * a block back into the list cannot fail) and return failure.
+ */
+
+static void *alloc_block(struct k_mem_pool *p, int l, size_t lsz)
+{
+	sys_dnode_t *block;
+	int key = irq_lock();
+
+	block = sys_dlist_get(&p->levels[l].free_list);
+	if (block) {
+		clear_free_bit(p, l, block_num(p, block, lsz));
+	}
+	irq_unlock(key);
+
+	return block;
+}
+
+static void free_block(struct k_mem_pool *p, int level, size_t *lsizes, int bn)
+{
+	int i, key, lsz = lsizes[level];
+	void *block = block_ptr(p, lsz, bn);
+
+	key = irq_lock();
+
+	set_free_bit(p, level, bn);
+
+	if (level && partner_bits(p, level, bn) == 0xf) {
+		for (i = 0; i < 4; i++) {
+			int b = (bn & ~3) + i;
+
+			clear_free_bit(p, level, b);
+			if (b != bn &&
+			    block_fits(p, block_ptr(p, lsz, b), lsz)) {
+				sys_dlist_remove(block_ptr(p, lsz, b));
+			}
+		}
+
+		irq_unlock(key);
+		free_block(p, level-1, lsizes, bn / 4); /* tail recursion! */
```

Comment:
Curious: Does the compiler really eliminates the call in this case?  I've seen it doing on cases where you return the result of calling a function recursively, transforming it into a JMP.